### PR TITLE
feat: add resources management validations

### PR DESCRIPTION
**Pull request by alexrodfe**
## Description

- Added new process resource limits with CPU and Memory configuration
- Added new validations for these configurations
- Added tests to support validations
- Changed coverpkg to enhance coverage

## Motivation and Context

Users will now be able to setup resource management, by deciding the base and limits of CPU and memory usage. These configurations are now available in the krt.yaml.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other changes (ci configuration, documentation or any other kind of changes)

## Checklist:

- [X] I have created tests for my code changes, and the tests are passing.
- [X] I have executed the pre-commit hooks locally.
- [ ] My change requires a change to the documentation (create a new issue if the documentation has not been updated).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/krt/krt_builder_test.go [hunk 1]
```diff
@@ -22,6 +22,14 @@ func NewKrtBuilder() *KrtBuilder {
 							Name:  "test-trigger",
 							Type:  krt.ProcessTypeTrigger,
 							Image: "test-trigger-image",
+							CPU: &krt.ProcessCPU{
```
Comment (kafkaphoenix) [inline pkg/krt/krt_builder_test.go:25]: in my examples i was using for memory Mi is M similar?

**Review by kafkaphoenix — DISMISSED** [review]

File: pkg/krt/krt_builder_test.go (hunk 1 above)
Comment (alexrodfe) [reply pkg/krt/krt_builder_test.go:25]: Kubernetes accepts both M and Mi, and all other measurements.
The difference between M and Mi is that M is x1000 and Mi is x1024 . More bit-wise correct (power of 2)

File: pkg/errors/errors.go [hunk 2]
```diff
@@ -26,6 +26,10 @@ var ErrInvalidWorkflowType = errors.New("invalid workflow type, must be either '
 var ErrInvalidProcessType = errors.New("invalid process type, must be either 'trigger', 'task' or 'exit'")
 var ErrInvalidProcessObjectStoreScope = errors.New("invalid process object store scope, must be either 'product' or 'workflow'")
 var ErrInvalidNetworkingProtocol = errors.New("invalid networking protocol, must be either 'UDP' or 'TCP'")
+var ErrInvalidProcessCPU = errors.New("invalid process CPU, must be of form '1', '0.5' or '100m'")
```
Comment (dfinteligenz) [inline pkg/errors/errors.go:29]: This should reference that the CPU value is for the limit somewhere.

Maybe change it to:
```suggestion
var ErrInvalidProcessCPULimits = errors.New("invalid process CPU limit value, must be of form '1', '0.5' or '100m'")
```

File: pkg/errors/errors.go [hunk 3]
```diff
@@ -26,6 +26,10 @@ var ErrInvalidWorkflowType = errors.New("invalid workflow type, must be either '
...
+var ErrInvalidProcessCPURelation = errors.New("invalid process CPU relation, 'limit' cannot be lower than 'request'")
```
Comment (dfinteligenz) [inline pkg/errors/errors.go:30]:
```suggestion
var ErrInvalidProcessCPURelation = errors.New("invalid process CPU, 'limit' cannot be lower than 'request'")
```

File: pkg/errors/errors.go [hunk 4]
```diff
@@ -26,6 +26,10 @@ var ErrInvalidWorkflowType = errors.New("invalid workflow type, must be either '
...
+var ErrInvalidProcessMemory = errors.New("invalid process memory, must be of form '350M' or '1Gi'")
```
Comment (dfinteligenz) [inline pkg/errors/errors.go:31]: Same as CPU, we should reference that this is a resource limit value somewhere in the message.

```suggestion
var ErrInvalidProcessMemory = errors.New("invalid process memory limit, must be of form '350M' or '1Gi'")
```

File: pkg/errors/errors.go [hunk 5]
```diff
@@ -26,6 +26,10 @@ var ErrInvalidWorkflowType = errors.New("invalid workflow type, must be either '
...
+var ErrInvalidProcessMemoryRelation = errors.New("invalid process memory relation, 'limit' cannot be lower than 'request'")
```
Comment (dfinteligenz) [inline pkg/errors/errors.go:32]:
```suggestion
var ErrInvalidProcessMemoryRelation = errors.New("invalid process memory, 'limit' cannot be lower than 'request'")
```

File: pkg/krt/krt.go [hunk 6]
```diff
@@ -124,3 +125,18 @@ func (np NetworkingProtocol) IsValid() bool {
...
+
+type ProcessMemory struct {
+	Request string `yaml:"request"`
+	Limit   string `yaml:"limit"`
+}
+
+type ProcessResourceLimits struct {
+	CPU    *ProcessCPU    `yaml:"CPU"`
+	Memory *ProcessMemory `yaml:"memory"`
+}
```
Comment (dfinteligenz) [inline pkg/krt/krt.go:142]: These two types could be replaced by a generic ResourceLimits struct
```suggestion
type ResourceLimit struct {
	Request string `yaml:"request"`
	Limit   string `yaml:"limit"`
}

type ProcessResourceLimits struct {
	CPU    *ResourceLimit    `yaml:"CPU"`
	Memory *ResourceLimit `yaml:"memory"`
}
```

**Review by dfinteligenz — CHANGES_REQUESTED** [review]

File: pkg/errors/errors.go (hunk 2 above)
Comment (alexrodfe) [reply pkg/errors/errors.go:29]: This error refers to both the limit and request variables

File: pkg/errors/errors.go (hunk 4 above)
Comment (alexrodfe) [reply pkg/errors/errors.go:31]: Ditto, refers to both values

File: pkg/krt/krt.go (hunk 6 above)
Comment (alexrodfe) [reply pkg/krt/krt.go:142]: Good catch

**Review by kafkaphoenix — APPROVED** [review]
Good work!

File: pkg/errors/errors.go (hunk 2 above)
Comment (dfinteligenz) [reply pkg/errors/errors.go:29]: Yep, but both limit and requests are resource limits

**Review by kafkaphoenix — APPROVED** [review]

**Review by dfinteligenz — DISMISSED** [review]

**Review by kafkaphoenix — APPROVED** [review]

**Pull request merged**